docs/deprecation.md

```diff
@@ -0,0 +1,23 @@
+# Deprecation Policy
+
+## User-facing commands and their arguments
+
+Whenever you rename a command or make a previously optional argument required,
+we require that you preserve the old command invocations keeps on working for 6
+months (so 6 releases, since we release monthly) with a deprecation message.
```

Comment:
(For discussion) This might be a mostly a stylistic difference, but I'd prefer to make this document softer.

For example, I'd say,

> When we rename commands or make previously optional arguments required, we usually try to keep the old command invocations working for 6 months...

(there are also a few unrelated fixups in the above)


We could also insert

> When we rename *commonly used* commands...

Reply:
Yes, IMO, these are more like rules of thumb than something we should strictly enforce. I think we should always balance how hard it is to keep compatibility vs how hard it is for users to adapt. I mentioned on Discord the example of file-level vs tree-level conflicts. In that case, dropping compatibility completely means that users have to re-clone their repo. As a result, we have been much more conservative with that deprecation. We still support file-level conflicts in old repos over a year later.

I think (but I may be wrong) when we originally started talking about the 6-month policy, it was about storage formats. I think it's fair to have a strict rule in that case saying that a repo created with a jj version less than 6 months old should be readable by the latest version.

Reply:
Yes, fair points. I have left the storage formats unmentioned since I had no good formulation of it.